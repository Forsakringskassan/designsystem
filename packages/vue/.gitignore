# vite dev
/src/vite-dev/app.vue
