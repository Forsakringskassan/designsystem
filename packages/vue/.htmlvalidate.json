{
    "extends": ["@fkui/vue:recommended"],
    "plugins": ["./htmlvalidate"],
    "transform": {
        "^.*\\.spec.ts$": "html-validate-vue:html"
    }
}
