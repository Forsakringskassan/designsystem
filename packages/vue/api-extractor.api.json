{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/api.d.ts",
    "dtsRollup": {
        "enabled": false
    },
    "messages": {
        "compilerMessageReporting": {
            "default": {
                "logLevel": "none"
            }
        }
    }
}
