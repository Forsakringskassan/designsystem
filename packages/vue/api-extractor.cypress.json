{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/cypress/index.d.ts",
    "compiler": {
        "tsconfigFilePath": "<projectFolder>/tsconfig.pageobjects.json"
    },
    "apiReport": {
        "reportFileName": "<unscopedPackageName>-cypress.api.md"
    },
    "docModel": {
        "enabled": true,
        "apiJsonFilePath": "<projectFolder>/temp/cypress.api.json"
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/cypress.d.ts"
    },
    "messages": {
        "extractorMessageReporting": {
            "ae-unresolved-link": {
                "logLevel": "none"
            }
        }
    }
}
