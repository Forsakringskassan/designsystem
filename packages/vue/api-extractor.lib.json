{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "compiler": {
        "tsconfigFilePath": "<projectFolder>/tsconfig.lib.json",
        "overrideTsconfig": {
            "moduleResolution": "node"
        }
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/index.d.ts"
    },
    "messages": {
        "compilerMessageReporting": {
            "default": {
                "logLevel": "none"
            }
        }
    }
}
