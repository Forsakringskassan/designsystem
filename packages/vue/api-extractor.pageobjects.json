{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/pageobjects.d.ts",
    "compiler": {
        "tsconfigFilePath": "<projectFolder>/tsconfig.pageobjects.json"
    },
    "apiReport": {
        "reportFileName": "<unscopedPackageName>-pageobjects.api.md"
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/pageobject.d.ts"
    }
}
