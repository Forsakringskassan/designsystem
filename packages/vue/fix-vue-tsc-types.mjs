import fs from "node:fs/promises";
import path from "node:path";
import { glob } from "glob";

let n = 0;
console.group("Patching vue-tsc dts files (microsoft/rushstack#5146)");

const filenames = await glob("**/*.vue.d.ts", { cwd: "temp/types" });
const promises = filenames.map(async (filename) => {
    const filePath = path.join("temp/types", filename);
    const content = await fs.readFile(filePath, "utf-8");
    const updated = content.replace(
        /declare var (__VLS_\d+)/,
        "declare const $1",
    );
    if (content !== updated) {
        await fs.writeFile(filePath, updated, "utf-8");
        n++;
        console.log(filename);
    }
});
await Promise.all(promises);

console.groupEnd();
console.log(`${n} file${n === 1 ? "" : "s"} patches\n`);
