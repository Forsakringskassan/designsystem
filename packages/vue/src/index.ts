export * from "./components";
export {
    config,
    type FKUIConfig,
    FKUIConfigButtonOrder,
    type FKUIContext,
    type MaybeWithFKUIContext,
    setRunningContext,
} from "./config";
export * from "./internal-components";
export * from "./plugins";
export * from "./types";
export * from "./utils";
