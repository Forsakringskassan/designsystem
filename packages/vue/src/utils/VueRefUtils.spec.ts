import { defineComponent } from "vue";
import { VueWrapper, mount } from "@vue/test-utils";
import {
    findElementFromVueRef,
    getElementFromVueRef,
    getElementsFromVueRef,
    getHTMLElementFromVueRef,
    getHTMLElementsFromVueRef,
    getSortedHTMLElementsFromVueRef,
    refIsElement,
    refIsElementArray,
    refIsHTMLElementArray,
    refIsVue,
    refIsVueArray,
} from "./VueRefUtils";

const REF_NAME = "myReference";
const TEMPLATE_WITHOUT_A_REF = /* HTML */ `
    <div><p>Lonely in the abyss.</p></div>
`;
const TEMPLATE_WITH_AN_ELEMENT_REF = /* HTML */ `
    <div><svg ref="${REF_NAME}">Find me!</svg></div>
`;
const TEMPLATE_WITH_AN_HTMLELEMENT_REF = /* HTML */ `
    <div><p ref="${REF_NAME}">Find me!</p></div>
`;
const TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF = /* HTML */ `
    <ul>
        <li v-for="item in ['1', '2', '3', '4']" ref="${REF_NAME}">
            {{ item }}
        </li>
    </ul>
`;
const TEMPLATE_WITH_AN_INDEXED_ELEMENT_ARRAY_REF = /* HTML */ `
    <ul>
        <li
            v-for="(item, index) in ['1', '2', '3', '4']"
            ref="${REF_NAME}"
            :data-ref-index="index"
        >
            {{ item }}
        </li>
    </ul>
`;
const TEMPLATE_WITH_A_COMPONENT_REF = /* HTML */ `
    <div><child-component ref="${REF_NAME}" message="Find me!" /></div>
`;
const TEMPLATE_WITH_A_COMPONENT_ARRAY_REF = /* HTML */ `
    <ul>
        <li v-for="item in ['1', '2', '3', '4']">
            <child-component :message="item" ref="${REF_NAME}" />
        </li>
    </ul>
`;

const ChildComponent = defineComponent({
    name: "ChildComponent",
    template: /* HTML */ ` <span>{{ message }}</span> `,
    props: {
        message: {
            type: String,
            default: "I have nothing to say.",
        },
    },
});

function createWrapper(template: string): VueWrapper {
    const TestComponent = defineComponent({
        name: "TestComponent",
        components: { ChildComponent },
        template,
    });

    return mount(TestComponent);
}

describe("refIsElement", () => {
    it.each`
        template                               | expected | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${false} | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${true}  | ${"ref is an Element"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${false} | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${false} | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${false} | ${"ref is an array of Vue components"}
    `("should return $expected if $description", ({ template, expected }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = refIsElement(ref);

        expect(result).toEqual(expected);
    });
});

describe("refIsElementArray", () => {
    it.each`
        template                               | expected | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${false} | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${false} | ${"ref is an Element"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${true}  | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${false} | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${false} | ${"ref is an array of Vue components"}
    `("should return $expected if $description", ({ template, expected }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = refIsElementArray(ref);

        expect(result).toEqual(expected);
    });
});

describe("refIsHTMLElementArray", () => {
    it.each`
        template                               | expected | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${false} | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${true}  | ${"ref is an HTMLElement array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${false} | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${false} | ${"ref is an array of Vue components"}
    `("should return $expected if $description", ({ template, expected }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = refIsHTMLElementArray(ref);

        expect(result).toEqual(expected);
    });
});

describe("refIsVue", () => {
    it.each`
        template                               | expected | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${false} | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${false} | ${"ref is an Element"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${false} | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${true}  | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${false} | ${"ref is an array of Vue components"}
    `("should return $expected if $description", ({ template, expected }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = refIsVue(ref);

        expect(result).toEqual(expected);
    });
});

describe("refIsVueArray", () => {
    it.each`
        template                               | expected | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${false} | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${false} | ${"ref is an Element"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${false} | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${false} | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${true}  | ${"ref is an array of Vue components"}
    `("should return $expected if $description", ({ template, expected }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = refIsVueArray(ref);

        expect(result).toEqual(expected);
    });
});

describe("getElementsFromVueRef", () => {
    it.each`
        template                               | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${"ref is an Element"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${"ref is an array of Vue components"}
    `("should match snapshot if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = getElementsFromVueRef(ref);

        expect(result).toMatchSnapshot();
    });
});

describe("getHTMLElementsFromVueRef", () => {
    it.each`
        template                               | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${"ref is an HTMLElement array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${"ref is an array of Vue components"}
    `("should match snapshot if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = getHTMLElementsFromVueRef(ref);

        expect(result).toMatchSnapshot();
    });
});

describe("getSortedHTMLElementsFromVueRef", () => {
    it("should return array with original order if ref is an unindexed HTMLElement array", () => {
        const wrapper = createWrapper(TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF);
        const ref = [...(wrapper.vm.$refs[REF_NAME] as unknown[])];
        const unorderedRef = [...ref].reverse();
        expect(unorderedRef).not.toEqual(ref);

        const result = getSortedHTMLElementsFromVueRef(unorderedRef);

        expect(result).toEqual(unorderedRef);
    });
    it("should return sorted array if ref is an indexed HTMLElement array", () => {
        const wrapper = createWrapper(
            TEMPLATE_WITH_AN_INDEXED_ELEMENT_ARRAY_REF,
        );
        const ref = [...(wrapper.vm.$refs[REF_NAME] as unknown[])];
        const unorderedRef = [...ref].reverse();
        expect(unorderedRef).not.toEqual(ref);

        const result = getSortedHTMLElementsFromVueRef(unorderedRef);

        expect(result).toEqual(ref);
    });
});

describe("findElementFromVueRef", () => {
    it.each`
        template                               | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${"ref is an Element"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}       | ${"ref is a Vue component"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${"ref is an array of Vue components"}
    `("should match snapshot if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = findElementFromVueRef(ref);

        expect(result).toMatchSnapshot();
    });
});

describe("getElementFromVueRef", () => {
    it.each`
        template                               | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${"ref is an array of Vue components"}
    `("should throw error if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];

        const call = (): void => {
            getElementFromVueRef(ref);
        };

        expect(call).toThrow();
    });

    it.each`
        template                         | description
        ${TEMPLATE_WITH_AN_ELEMENT_REF}  | ${"ref is an Element"}
        ${TEMPLATE_WITH_A_COMPONENT_REF} | ${"ref is a Vue component"}
    `("should match snapshot if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = getElementFromVueRef(ref);

        expect(result).toMatchSnapshot();
    });
});

describe("getHTMLElementFromVueRef", () => {
    it.each`
        template                               | description
        ${TEMPLATE_WITHOUT_A_REF}              | ${"ref does not exist"}
        ${TEMPLATE_WITH_AN_ELEMENT_ARRAY_REF}  | ${"ref is an Element array"}
        ${TEMPLATE_WITH_A_COMPONENT_ARRAY_REF} | ${"ref is an array of Vue components"}
        ${TEMPLATE_WITH_AN_ELEMENT_REF}        | ${"ref is an array of Vue components"}
    `("should throw error if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];

        const call = (): void => {
            getHTMLElementFromVueRef(ref);
        };

        expect(call).toThrow();
    });

    it.each`
        template                            | description
        ${TEMPLATE_WITH_AN_HTMLELEMENT_REF} | ${"ref is an Element"}
        ${TEMPLATE_WITH_A_COMPONENT_REF}    | ${"ref is a Vue component"}
    `("should match snapshot if $description", ({ template }) => {
        const wrapper = createWrapper(template);

        const ref = wrapper.vm.$refs[REF_NAME];
        const result = getHTMLElementFromVueRef(ref);

        expect(result).toMatchSnapshot();
    });
});
