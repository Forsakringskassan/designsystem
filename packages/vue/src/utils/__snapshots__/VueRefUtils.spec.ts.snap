// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findElementFromVueRef should match snapshot if ref does not exist 1`] = `undefined`;

exports[`findElementFromVueRef should match snapshot if ref is a Vue component 1`] = `
<span>
  Find me!
</span>
`;

exports[`findElementFromVueRef should match snapshot if ref is an Element 1`] = `
<svg>
  Find me!
</svg>
`;

exports[`findElementFromVueRef should match snapshot if ref is an Element array 1`] = `undefined`;

exports[`findElementFromVueRef should match snapshot if ref is an array of Vue components 1`] = `undefined`;

exports[`getElementFromVueRef should match snapshot if ref is a Vue component 1`] = `
<span>
  Find me!
</span>
`;

exports[`getElementFromVueRef should match snapshot if ref is an Element 1`] = `
<svg>
  Find me!
</svg>
`;

exports[`getElementsFromVueRef should match snapshot if ref does not exist 1`] = `[]`;

exports[`getElementsFromVueRef should match snapshot if ref is a Vue component 1`] = `
[
  <span>
    Find me!
  </span>,
]
`;

exports[`getElementsFromVueRef should match snapshot if ref is an Element 1`] = `
[
  <svg>
    Find me!
  </svg>,
]
`;

exports[`getElementsFromVueRef should match snapshot if ref is an Element array 1`] = `
[
  <li>
    1
  </li>,
  <li>
    2
  </li>,
  <li>
    3
  </li>,
  <li>
    4
  </li>,
]
`;

exports[`getElementsFromVueRef should match snapshot if ref is an array of Vue components 1`] = `
[
  <span>
    1
  </span>,
  <span>
    2
  </span>,
  <span>
    3
  </span>,
  <span>
    4
  </span>,
]
`;

exports[`getHTMLElementFromVueRef should match snapshot if ref is a Vue component 1`] = `
<span>
  Find me!
</span>
`;

exports[`getHTMLElementFromVueRef should match snapshot if ref is an Element 1`] = `
<p>
  Find me!
</p>
`;

exports[`getHTMLElementsFromVueRef should match snapshot if ref does not exist 1`] = `[]`;

exports[`getHTMLElementsFromVueRef should match snapshot if ref is a Vue component 1`] = `
[
  <span>
    Find me!
  </span>,
]
`;

exports[`getHTMLElementsFromVueRef should match snapshot if ref is an HTMLElement array 1`] = `
[
  <li>
    1
  </li>,
  <li>
    2
  </li>,
  <li>
    3
  </li>,
  <li>
    4
  </li>,
]
`;

exports[`getHTMLElementsFromVueRef should match snapshot if ref is an array of Vue components 1`] = `
[
  <span>
    1
  </span>,
  <span>
    2
  </span>,
  <span>
    3
  </span>,
  <span>
    4
  </span>,
]
`;
